File: tree/leetcode_199/solution.py
# Definition for a binary tree node.
# import pdb
import collections
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def rightSideView(self, root: [TreeNode]) -> list[int]:
        if not root:
            return []
        
        queue = collections.deque()
        queue.append(root)
        res = []
        while queue:
            len_que = len(queue)
            rightest_val = None

            for i in range(len_que):
                node = queue.popleft()
                # pdb.set_trace()
                if node:
                    rightest_val = node.val
                    queue.append(node.left)
                    queue.append(node.right)

                if i == len_que - 1 and rightest_val is not None:
                    res.append(rightest_val)

        return res

File: tree/leetcode_199/solution.rb
require "../builder"
# @param {TreeNode} root
# @return {Integer[]}
def right_side_view(root)
    if root.nil?
        return []
    end

    if !root.left and !root.right
        return [root.val]
    end

    queue = [root]
    result = []
    while queue.length > 0
        node_count = queue.length
        for i in 1..node_count
            node = queue.shift
            if i == node_count
                result.append(node.val)
            end
            if node.left
                queue.append(node.left)
            end
            if node.right
                queue.append(node.right)
            end
        end
    end
    return result
end



test_cases = [
    {
        root: [1,2,3,nil,5,nil,4],
        expect: [1,3,4]
    },
    {
        root: [1,nil,3],
        expect: [1,3]
    },
    {
        root: [],
        expect: []
    },
    {
        root: [1,2],
        expect: [1,2]
    },
]

builder = BuilderTreeNode.new()
for test_case in test_cases
    root = builder.build(test_case[:root])
    result = right_side_view(root)

    if result == test_case[:expect]
        print("Pass!\n")
    else
        print("Not pass, result: #{result}\n")
    end
end